File: src/main/java/ru/ezhov/rocket/action/infrastructure/YmlRocketActionSettingsRepository.java
package ru.ezhov.rocket.action.infrastructure;

import org.yaml.snakeyaml.Yaml;
import ru.ezhov.rocket.action.RocketActionSettingsRepository;
import ru.ezhov.rocket.action.RocketActionSettingsRepositoryException;
import ru.ezhov.rocket.action.api.RocketActionSettings;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.UUID;

public class YmlRocketActionSettingsRepository implements RocketActionSettingsRepository {

    private final URI uri;
    private static final String TYPE = "type";
    private static final String ID = "_id";
    private static final String ACTIONS = "actions";

    public YmlRocketActionSettingsRepository(URI uri) {
        this.uri = uri;
    }

    @Override
    public List<RocketActionSettings> actions() throws RocketActionSettingsRepositoryException {
        try (InputStream inputStream = uri.toURL().openStream()) {
            List<RocketActionSettings> actions = new ArrayList<>();
            Yaml yaml = new Yaml();
            Map<String, Object> obj = yaml.load(inputStream);

            for (Map.Entry<String, Object> e : obj.entrySet()) {
                if (ACTIONS.equals(e.getKey())) {
                    ArrayList<LinkedHashMap<String, Object>> linkedHashMaps =
                            (ArrayList<LinkedHashMap<String, Object>>) e.getValue();
                    for (LinkedHashMap<String, Object> l : linkedHashMaps) {
                        actions.add(createAction(l));
                    }
                }
            }
            return actions;
        } catch (Exception ex) {
            throw new RocketActionSettingsRepositoryException("//TODO", ex);
        }
    }

    @Override
    public void save(List<RocketActionSettings> settings) throws RocketActionSettingsRepositoryException {
        File file = new File(uri.getPath());
        try (OutputStreamWriter outputStreamWriter =
                     new OutputStreamWriter(
                             new FileOutputStream(file),
                             StandardCharsets.UTF_8)
        ) {
            Yaml yaml = new Yaml();

            List<Map<String, Object>> all = new ArrayList<>();
            recursiveSettings(settings, all);

            Map<String, Object> map = new LinkedHashMap<>();
            map.put(ACTIONS, all);
            yaml.dump(map, outputStreamWriter);
        } catch (Exception ex) {
            throw new RocketActionSettingsRepositoryException("//TODO", ex);
        }
    }

    private void recursiveSettings(List<RocketActionSettings> settings, List<Map<String, Object>> actions) {
        for (RocketActionSettings data : settings) {
            Map<String, Object> object = new LinkedHashMap<>();
            object.put(TYPE, data.type());
            object.put(ID, data.id());
            data.settings().forEach(object::put);
            final List<RocketActionSettings> actionsOriginal = data.actions();
            if (!actionsOriginal.isEmpty()) {
                List<Map<String, Object>> actionsForWrite = new ArrayList<>();
                recursiveSettings(actionsOriginal, actionsForWrite);
                object.put(ACTIONS, actionsForWrite);
            }

            actions.add(object);
        }
    }

    private RocketActionSettings createAction(LinkedHashMap<String, Object> action) throws RocketActionSettingsRepositoryException {
        return QuickActionFactory.create(
                getOrGenerateId(action),
                action.get(TYPE).toString(),
                action
        );
    }

    private static String getOrGenerateId(LinkedHashMap<String, Object> action) {
        Object idAsObject = action.get(ID);
        if (idAsObject == null || "".equals(idAsObject.toString())) {
            idAsObject = UUID.randomUUID().toString();
        }
        return idAsObject.toString();
    }

    private static class QuickActionFactory {
        static RocketActionSettings createAction(LinkedHashMap<String, Object> action) throws RocketActionSettingsRepositoryException {
            return QuickActionFactory.create(
                    getOrGenerateId(action),
                    action.get(TYPE).toString(),
                    action
            );
        }

        static RocketActionSettings create(String id, String actionType, LinkedHashMap<String, Object> action) throws RocketActionSettingsRepositoryException {
            ArrayList<LinkedHashMap<String, Object>> actions =
                    (ArrayList<LinkedHashMap<String, Object>>) action.get(ACTIONS);
            action.remove(TYPE);
            action.remove(ID);
            action.remove(ACTIONS);

            if (actions == null || actions.isEmpty()) {
                Map<String, String> map = new TreeMap<>();
                action.forEach((k, v) -> map.put(k, v == null ? "" : v.toString()));

                return new MutableRocketActionSettings(id, actionType, map, Collections.emptyList());
            } else {
                List<RocketActionSettings> settings = new ArrayList<>();
                for (LinkedHashMap<String, Object> a : actions) {
                    settings.add(createAction(a));
                }
                Map<String, String> map = new TreeMap<>();
                action.forEach((k, v) -> map.put(k, v.toString()));
                return new MutableRocketActionSettings(id, actionType, map, settings);
            }
        }
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/infrastructure/YmlRocketActionSettingsRepository.kt
package ru.ezhov.rocket.action.infrastructure

import org.yaml.snakeyaml.Yaml
import ru.ezhov.rocket.action.domain.RocketActionSettingsRepository
import ru.ezhov.rocket.action.domain.RocketActionSettingsRepositoryException
import ru.ezhov.rocket.action.api.RocketActionSettings
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.net.URI
import java.nio.charset.StandardCharsets
import java.util.*

class YmlRocketActionSettingsRepository(private val uri: URI) : RocketActionSettingsRepository {
    @Throws(RocketActionSettingsRepositoryException::class)
    override fun actions(): List<RocketActionSettings> {
        try {
            uri.toURL().openStream().use { inputStream ->
                val actions: MutableList<RocketActionSettings> = ArrayList()
                val yaml = Yaml()
                val obj = yaml.load<Map<String, Any>>(inputStream)
                for ((key, value) in obj) {
                    if (ACTIONS == key) {
                        val linkedHashMaps = value as ArrayList<LinkedHashMap<String, Any>>
                        for (l in linkedHashMaps) {
                            actions.add(createAction(l))
                        }
                    }
                }
                return actions
            }
        } catch (ex: Exception) {
            throw RocketActionSettingsRepositoryException("//TODO", ex)
        }
    }

    @Throws(RocketActionSettingsRepositoryException::class)
    override fun save(settings: List<RocketActionSettings>) {
        val file = File(uri!!.path)
        try {
            OutputStreamWriter(
                    FileOutputStream(file),
                    StandardCharsets.UTF_8).use { outputStreamWriter ->
                val yaml = Yaml()
                val all: MutableList<Map<String?, Any?>> = ArrayList()
                recursiveSettings(settings, all)
                val map: MutableMap<String, Any> = LinkedHashMap()
                map[ACTIONS] = all
                yaml.dump(map, outputStreamWriter)
            }
        } catch (ex: Exception) {
            throw RocketActionSettingsRepositoryException("//TODO", ex)
        }
    }

    private fun recursiveSettings(settings: List<RocketActionSettings>?, actions: MutableList<Map<String?, Any?>>) {
        for (data in settings!!) {
            val `object`: MutableMap<String?, Any?> = LinkedHashMap()
            `object`[TYPE] = data.type()
            `object`[ID] = data.id()
            data.settings().forEach { (key: String?, value: String?) -> `object`[key] = value }
            val actionsOriginal = data.actions()
            if (!actionsOriginal!!.isEmpty()) {
                val actionsForWrite: MutableList<Map<String?, Any?>> = ArrayList()
                recursiveSettings(actionsOriginal, actionsForWrite)
                `object`[ACTIONS] = actionsForWrite
            }
            actions.add(`object`)
        }
    }

    @Throws(RocketActionSettingsRepositoryException::class)
    private fun createAction(action: LinkedHashMap<String, Any>): RocketActionSettings {
        return QuickActionFactory.create(
                getOrGenerateId(action),
                action[TYPE].toString(),
                action
        )
    }

    private object QuickActionFactory {
        @kotlin.jvm.JvmStatic
        @Throws(RocketActionSettingsRepositoryException::class)
        fun createAction(action: LinkedHashMap<String, Any>): RocketActionSettings {
            return create(
                    getOrGenerateId(action),
                    action[TYPE].toString(),
                    action
            )
        }

        @kotlin.jvm.JvmStatic
        @Throws(RocketActionSettingsRepositoryException::class)
        fun create(id: String, actionType: String, action: LinkedHashMap<String, Any>): RocketActionSettings {
            val actions = action[ACTIONS] as ArrayList<LinkedHashMap<String, Any>>?
            action.remove(TYPE)
            action.remove(ID)
            action.remove(ACTIONS)
            return if (actions == null || actions.isEmpty()) {
                val map: MutableMap<String, String> = TreeMap()
                action.forEach { (k: String, v: Any?) -> map[k] = v?.toString() ?: "" }
                MutableRocketActionSettings(id, actionType, map, mutableListOf())
            } else {
                val settings: MutableList<RocketActionSettings> = ArrayList()
                for (a in actions) {
                    settings.add(createAction(a))
                }
                val map: MutableMap<String, String> = TreeMap()
                action.forEach { (k: String, v: Any) -> map[k] = v.toString() }
                MutableRocketActionSettings(id, actionType, map, settings)
            }
        }
    }

    companion object {
        private const val TYPE = "type"
        private const val ID = "_id"
        private const val ACTIONS = "actions"
        private fun getOrGenerateId(action: LinkedHashMap<String, Any>): String {
            var idAsObject = action[ID]
            if (idAsObject == null || "" == idAsObject.toString()) {
                idAsObject = UUID.randomUUID().toString()
            }
            return idAsObject.toString()
        }
    }
}